src/e3/pytest.py

```diff
@@ -0,0 +1,223 @@
+
+    return pytest.fixture(wrapper)
+
+
+def pytest_addoption(
+    parser: pytest.Parser, pluginmanager: pytest.PytestPluginManager
+) -> None:
+    group = parser.getgroup("e3")
+    group.addoption("--e3", action="store_true", help="Use e3 fixtures and reporting")
+    group.addoption("--e3-cov-rewrite", nargs=2, help="Use e3 fixtures and reporting")
+
+
+@pytest.fixture(autouse=True)
+def env_protect(request: pytest.FixtureRequest) -> None:
+    """Protection against environment change.
+
+    The fixture is enabled for all tests and does the following:
+
+    * store/restore env between each tests
+    * create a temporary directory and do a cd to it before each
+      test. The directory is automatically removed when test ends
+    """
+    if request.config.getoption("e3"):
+        Env().store()
+        tempd = mkdtemp()
+        cd(tempd)
+        Config.data = {}
+
+        os.environ["TZ"] = "UTC"
+        os.environ["E3_ENABLE_FEATURE"] = ""
+        os.environ["E3_CONFIG"] = "/dev/null"
+        if "E3_HOSTNAME" in os.environ:
+            del os.environ["E3_HOSTNAME"]
+
+        def restore_env() -> None:
+            Env().restore()
+            rm(tempd, True)
+
+        e3.log.activate(level=logging.DEBUG, e3_debug=True)
+        request.addfinalizer(restore_env)
+
+
+@pytest.hookimpl(trylast=True)
+def pytest_sessionfinish(session: pytest.Session, exitstatus: int) -> None:
+    """Manage the exit code depending on if errors were detected or not."""
+    if not session.config.getoption("e3"):
+        return
+    global test_errors
+    if test_errors:
+        # Return with an exit code of `3` if we encountered errors (not failures).
+        # This is the exit code that corresponds to an "internal error" according to the
+        # pytest docs, which is the closest thing to having an actual Python error in
+        # test code.
+        session.exitstatus = 3
+
+    if session.config.getoption("cov_source"):
+        cov_file = str(session.config.rootpath / ".coverage")
+        if session.config.getoption("e3_cov_rewrite"):
+            origin_dir, new_dir = session.config.getoption("e3_cov_rewrite")
+            fix_coverage_paths(origin_dir=origin_dir, new_dir=new_dir, cov_db=cov_file)
+
+        os_name = Env().build.os.name
+
+        cov = Coverage(data_file=cov_file)
+
+        # Exclude all lines matching the DEFAULT_EXCLUDE_LIST
+        for regex in DEFAULT_EXCLUDE_LIST:
+            cov.exclude(regex)
+
+        # Exclude all <os>-only line (unless that's for this OS)
+        for regex in (
+            "%s-only" % o
+            for o in ("darwin", "linux", "solaris", "windows", "bsd", "aix")
+            if o != os_name
+        ):
+            cov.exclude(regex)
+
+        # Handling of Windows
+        if os_name == "windows":
+            cov.exclude('if sys.platform != "win32":')
+        else:
+            cov.exclude('if sys.platform == "win32":')
+            cov.exclude("unix: no cover")
```

Comment:
Works for Darwin too (I guess you tested it, but just to make sure ...)

Reply:
I don't get your comment. If this either to activate coverage on windows only, or activate coverage on other platforms.

Reply:
Yes, macOS and Linux are considered unix here.